oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java

```diff
@@ -137,42 +135,89 @@
+			Map<String, Object> configuration = rest.exchange(request, typeReference).getBody();
+			AuthorizationServerMetadata metadata = parse(configuration, AuthorizationServerMetadata::parse);
+			validateJwksUri(metadata);
+			ClientRegistration.Builder builder = withProviderConfiguration(metadata, issuer.toASCIIString())
+					.jwkSetUri(metadata.getJWKSetURI().toASCIIString());
+			String userinfoEndpoint = (String) configuration.get("userinfo_endpoint");
+			if (userinfoEndpoint != null) {
+				builder.userInfoUri(userinfoEndpoint);
+			}
+			return builder;
+		};
 	}
 
-	private static URI oauth(URI issuer) {
-		return UriComponentsBuilder.fromUri(issuer)
+	private static Supplier<ClientRegistration.Builder> oauth(URI issuer) {
+		URI uri = UriComponentsBuilder.fromUri(issuer)
 				.replacePath(OAUTH_METADATA_PATH + issuer.getPath()).build(Collections.emptyMap());
+		return () -> {
+			RequestEntity<Void> request = RequestEntity.get(uri).build();
+			Map<String, Object> configuration = rest.exchange(request, typeReference).getBody();
+			AuthorizationServerMetadata metadata = parse(configuration, AuthorizationServerMetadata::parse);
+			ClientRegistration.Builder builder = withProviderConfiguration(metadata, issuer.toASCIIString())
+					.jwkSetUri(metadata.getJWKSetURI() != null ? metadata.getJWKSetURI().toASCIIString() : null);
+			String userinfoEndpoint = (String) configuration.get("userinfo_endpoint");
+			if (userinfoEndpoint != null) {
+				builder.userInfoUri(userinfoEndpoint);
+			}
+			return builder;
+		};
+	}
+
+	/**
+	 * As per the <a href="https://openid.net/specs/openid-connect-discovery-1_0.html">
+	 * The OpenID Connect Discovery 1.0</a> specification, the JWK set URI is a required
+	 * field in <a href="https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationResponse">
+	 * OpenID Provider Configuration Response</a>, however, in the <a href="https://tools.ietf.org/html/rfc8414">
+	 * RFC 8414</a> specification it is OPTIONAL in <a href="https://tools.ietf.org/html/rfc8414#section-2">
+	 * Authorization Server Metadata</a> response for OAuth2.
+	 *
+	 * @param metadata
+	 */
+	private static void validateJwksUri(AuthorizationServerMetadata metadata) {
+		if (metadata.getJWKSetURI() == null) {
+			throw new IllegalArgumentException("The public JWK set URI must not be null");
+		}
 	}
 
-	private static Map<String, Object> getConfiguration(String issuer, URI... uris) {
+	@SafeVarargs
+	private static ClientRegistration.Builder getBuilder(String issuer, Supplier<ClientRegistration.Builder>... suppliers) {
 		String errorMessage = "Unable to resolve Configuration with the provided Issuer of \"" + issuer + "\"";
-		for (URI uri : uris) {
+		for (Supplier<ClientRegistration.Builder> supplier : suppliers) {
 			try {
-				RequestEntity<Void> request = RequestEntity.get(uri).build();
-				return rest.exchange(request, typeReference).getBody();
+				ClientRegistration.Builder builder = supplier.get();
```

Comment:
We could probably simplify this a bit:

```suggestion
				return supplier.get();
```

Reply:
👍